pants

```diff
@@ -73,23 +73,54 @@ EOF
 
 # The high-level flow:
 # 1.) Resolve the Python interpreter, first reading from the env var $PYTHON,
-#     then reading from pants.ini, then defaulting to `python2.7`.
+#     then reading from pants.ini, then determining the default based off of
+#     which Python versions the current `pants_version` supports.
 # 2.) Grab pants version from pants.ini or default to latest.
 # 3.) Check for a venv via a naming/path convention and execute if found.
 # 4.) Otherwise create venv and re-exec self.
 #
 # After that pants itself will handle making sure any requested plugins
 # are installed and up to date.
 
-function determine_pants_runtime_python_version {
+function determine_default_python_exe {
+  pants_version="$(get_pants_ini_config_value 'pants_version')"
+  # If `pants_version` not pinned in `pants.ini`, default to what the most
+  # recent stable release supports.
+  if [[ -z "${pants_version}" ]]; then
+    supported_python_versions=('3.6' '2.7')
+    supported_message='2.7 or 3.6'
+  else
+    pants_minor_version="$(echo "${pants_version}" | cut -d '.' -f2)"
+    if [[ "${pants_minor_version}" -ge 17 ]]; then
+      supported_python_versions=('3.6' '3.7' '3.8' '3')
```

Comment:
This is not great. Ideally we'd just accept any python3.* And then use it to run a test snippet that checked `sys.version_info[:2] >= (3, 6) and sys.version_info[:2] < (4, 0)`. Then maybe sort the answers in the affirmative and pick the lowest. The main point being, do some slower stuff here to pick a 3 series interpreter if it avoids having to thrash this script againt support 3.9, etc. Currently we can't though since finding python is in the mainline of execution even after bootstrap :/. FWICT that is un-necessary. If we bootstrap pants version x.y.z, it's bootstrap path need not say which python if once the right choice for a given pants version, always the right choice. From my phone so I may not have taken in all the right context, but I think this makes sense.

Reply:
I'm a little confused how this would work. If there are multiple `python3`s when running `which -a python3`, how would we execute them all? Would we do something like this?

```
possible_pythons = ()
for bin in which -a python3
  version_info = sys.version_info[0:2]
  possible_pythons += (bin version_info)  # i.e. a tuple
done
find min from possible_pythons
```

Seems difficult to implement in Bash.

--

Alternatively, it's been stated somewhere that there won't be a 3.10 because Guido doesn't like two-digit minor versions. We could simply add `3.9` to the list and call it a day? Nothing bad will happen if `3.9` is not released yet, and we aren't claiming to support 4.

Is that fine? I can find the quote need be.

Reply:
I'm fine with handling 3.9 + quote. It would be nice to take interpreter selection out of the mainline still, but that could be a seperate change.